sdk/include/opentelemetry/sdk/trace/id_generator.h

```diff
@@ -0,0 +1,49 @@
+// Copyright 2021, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include "opentelemetry/trace/span_id.h"
+#include "opentelemetry/trace/trace_id.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+
+/** IdGenerator provides an interface for generating Trace Id and Span Id */
+class IdGenerator
+{
+
+public:
+  /** Returns a SpanId represented by opaque 128-bit trace identifier */
+  virtual opentelemetry::trace::SpanId GenerateSpanId() noexcept = 0;
```

Comment:
Should the contract enforce constness or not?

I can foresee an implementation that does mutate internal state, a simple implementation being an incremental counter that generates an ID based on the counter. But I'm not sure whether this is by design or not?

It seems Sampler likewise doesn't use `const`

Reply:
I agree. We shouldn't use `const`, as we can't assume the `IdGenerator` to be stateless.

Reply:
We shouldn't enforce constness, which is how `IdGenerator` is currently implemented. Let me know if you foresee any further changes?